### DISCUSSION TITLE
Enabling R2 (Cloudflare's S3-compatible object storage) for Mastodon's media backend

### DISCUSSION THREAD

**Discussion by ferdinandosimonetti** [mastodon/mastodon#20952] [Q&A] [answered]
### Pitch

Adding Cloudflare's R2 to the list of object storage provider.

### Motivation

Expanding Mastodon's S3 backend compatibility to include Cloudflare's offer could reduce storage-related costs for several instances (given R2's "no egress fees" policy and ample free GET requests' plafond),

**Comment by bearice** [comment, marked as answer]
R2 is S3 API compatible. My instance is using R2 currently.
Just set these environment varibles:
```
S3_ENABLED = "true"
S3_BUCKET = "mastodon"
S3_ENDPOINT = "https://<some-id>.r2.cloudflarestorage.com"
S3_ALIAS_HOST = "example.com" 
S3_PERMISSION = "private"
AWS_ACCESS_KEY_ID = "get one from cloudflare console"
AWS_SECRET_ACCESS_KEY = "get one from cloudflare console"
```
the only tricky part is `S3_PERMISSION = "private"` where has to be defined and ignored by R2 but default value "public" will cause R2 to throw an "ACL not implemented" error.

**Comment by ferdinandosimonetti** [comment]
Thanks a lot!

**Comment by umarhadi** [comment]
> R2 is S3 API compatible. My instance is using R2 currently. Just set these environment varibles:
> 
> ```
> S3_ENABLED = "true"
> S3_BUCKET = "mastodon"
> S3_ENDPOINT = "https://<some-id>.r2.cloudflarestorage.com"
> S3_ALIAS_HOST = "example.com" 
> S3_PERMISSION = "private"
> AWS_ACCESS_KEY_ID = "get one from cloudflare console"
> AWS_SECRET_ACCESS_KEY = "get one from cloudflare console"
> ```
> 
> the only tricky part is `S3_PERMISSION = "private"` where has to be defined and ignored by R2 but default value "public" will cause R2 to throw an "ACL not implemented" error.

thank youu!

**Comment by despens** [comment]
The [Mastodon documentation](https://docs.joinmastodon.org/admin/config/#cdn) states:

> You must serve the files with CORS headers, otherwise some functions of Mastodon's web UI will not work. For example, `Access-Control-Allow-Origin: *`

I wasn't able to get CORS headers working on R2.

@bearice did you notice any downside in Mastodon when this header missing with images etc that are served via R2?

(Of course I could use a reverse proxy to the bucket that adds the CORS header, but that kind of defeats the purpose of using Cloudflare…)

**Reply by bearice** [reply]
I had no problem so far, both with web UI and clients.
I think the reason is that cors checks don't apply on img tags and for most cases this is what r2 is used for.

**Reply by bearice** [reply]
BTW: there is a way to set the CORS headers, but not well-documented.
refs: https://kian.org.uk/configuring-cors-on-cloudflare-r2/

**Reply by despens** [reply]
I followed this tutorial but it didn't lead to the CORS headers being set for bucket items. I believe Cloudflare is just not there with the S3 API compatibility at the moment.

**Reply by bearice** [reply]
I can confirm it is working:
```
$ http OPTIONS https://i.moew.app/accounts/headers/109/309/158/711/867/083/original/fbd29ce031328d7e.png Origin:test.com
HTTP/1.1 204 No Content
Access-Control-Allow-Methods: GET, HEAD
Access-Control-Allow-Origin: *
CF-Cache-Status: DYNAMIC
CF-RAY: 76c1976648a17ecc-LAX
Connection: keep-alive
Server: cloudflare
Vary: Origin, Accept-Encoding
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
```

I think you'd check if you are use `PUT` not `POST` when creating CORS rules, you can also use `GET` to check it (yeah I made this mistake on my first try)

**Reply by despens** [reply]
Yes you are correct @bearice! I just did my test requests wrongly. Thank you for sharing the correct way to do it via the OPTIONS request.

**Reply by glamoroussky** [reply]
> I can confirm it is working:
> 
> ```
> $ http OPTIONS https://i.moew.app/accounts/headers/109/309/158/711/867/083/original/fbd29ce031328d7e.png Origin:test.com
> HTTP/1.1 204 No Content
> Access-Control-Allow-Methods: GET, HEAD
> Access-Control-Allow-Origin: *
> CF-Cache-Status: DYNAMIC
> CF-RAY: 76c1976648a17ecc-LAX
> Connection: keep-alive
> Server: cloudflare
> Vary: Origin, Accept-Encoding
> alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
> ```
> 
> I think you'd check if you are use `PUT` not `POST` when creating CORS rules, you can also use `GET` to check it (yeah I made this mistake on my first try)

Hi Could you please share a detail sample of the CORS policy, I was so messed up by this setting, thank you so much.

**Reply by despens** [reply]
Hey @glamoroussky, this is the Cloudflare CORS policy that I can confirm works for me:

```JSON
[
  {
    "AllowedOrigins": [
      "*"
    ],
    "AllowedMethods": [
      "GET",
      "HEAD"
    ]
  }
]
```

**Comment by glamoroussky** [comment]
Hi @despens , thank you so much. It still doesn't work for me. I have added a subdomain to the bucket from R2 setting.
Do I still need to configure a Nginx config for the subdomain?

**Reply by despens** [reply]
Hey @glamoroussky, you wouldn't need to configure nginx if you use the Cloudflare R2 bucket with a custom subdomain. Instead you need to set `S3_ALIAS_HOST` in the production environment to the subdomain you chose for the bucket. Proxying it through nginx is an unnecessary extra step I guess.

**Reply by glamoroussky** [reply]
@despens Hi mate, thanks for the clarification, I just found out the reason it doesn't work is I didn't add the new ips to Cloudflare's token....it works perfectly right now. Thank you so much.
